openvdb/tools/LevelSetMeasure.h

```diff
@@ -162,77 +156,122 @@ class LevelSetMeasure
     bool checkInterrupter();
 
-    using LeafT = typename TreeType::LeafNodeType;
-    using VoxelCIterT = typename LeafT::ValueOnCIter;
-    using LeafRange = typename ManagerType::LeafRange;
-    using LeafIterT = typename LeafRange::Iterator;
-
-    struct Measure2
+    struct MeasureArea
     {
-        Measure2(LevelSetMeasure* parent) : mParent(parent), mAcc(*mParent->mTree)
+        MeasureArea(LevelSetMeasure* parent) : mParent(parent), mStencil(*mParent->mGrid)
         {
+            if (parent->mInterrupter) parent->mInterrupter->start("Measuring area and volume of level set");
             if (parent->mGrainSize>0) {
                 tbb::parallel_for(parent->mLeafs->leafRange(parent->mGrainSize), *this);
             } else {
                 (*this)(parent->mLeafs->leafRange());
             }
+            std::thread t1([&](){parent->mArea   = parent->reduce(0);});
+            std::thread t2([&](){parent->mVolume = parent->reduce(1)/3.0;});
+            t1.join();
+            t2.join();
+            parent->mUpdateArea = false;
+            if (parent->mInterrupter) parent->mInterrupter->end();
         }
-        Measure2(const Measure2& other) : mParent(other.mParent), mAcc(*mParent->mTree) {}
+        MeasureArea(const MeasureArea& other) : mParent(other.mParent), mStencil(*mParent->mGrid) {}
         void operator()(const LeafRange& range) const;
         LevelSetMeasure* mParent;
-        typename GridT::ConstAccessor mAcc;
-    };
-    struct Measure3
+        mutable math::GradStencil<GridT, false> mStencil;
+    };// MeasureArea
+
+    struct MeasureCurvatures
     {
-        Measure3(LevelSetMeasure* parent) : mParent(parent), mAcc(*mParent->mTree)
+        MeasureCurvatures(LevelSetMeasure* parent) : mParent(parent), mStencil(*mParent->mGrid)
         {
+            if (parent->mInterrupter) parent->mInterrupter->start("Measuring curvatures of level set");
             if (parent->mGrainSize>0) {
                 tbb::parallel_for(parent->mLeafs->leafRange(parent->mGrainSize), *this);
             } else {
                 (*this)(parent->mLeafs->leafRange());
             }
+            std::thread t1([&](){parent->mTotMeanCurvature = parent->reduce(0);});
+            std::thread t2([&](){parent->mTotGausCurvature = parent->reduce(1);});
+            t1.join();
+            t2.join();
+            parent->mUpdateCurvature = false;
+            if (parent->mInterrupter) parent->mInterrupter->end();
         }
-        Measure3(const Measure3& other) : mParent(other.mParent), mAcc(*mParent->mTree) {}
+        MeasureCurvatures(const MeasureCurvatures& other) : mParent(other.mParent), mStencil(*mParent->mGrid) {}
         void operator()(const LeafRange& range) const;
         LevelSetMeasure* mParent;
-        typename GridT::ConstAccessor mAcc;
-    };
-    inline double reduce(double* first, double scale)
+        mutable math::CurvatureStencil<GridT, false> mStencil;
```

Comment:
minor: you could opt to move the member mStencil into the operator of these mini struct to avoid have to explicitly define the copy constructor

Reply:
That will imply that a new instance of the stencil is constructed each time the operator method is called. If on the other hand it's a member a stencil is only created for each new instance of the struct. Since the stencil is internally caching grid nodes it's not exactly lightweight so I want to re-use it as much as possible (without actually resorting to TLS)

Reply:
> That will imply that a new instance of the stencil is constructed each time the operator method is called. 

Isn't this also the case currently due the copy constructor of these structs re-creating it from the grid? I assume this is to *avoid* the copying of its cache unnecessarily. As tbb will copy this object for each subrange, isn't this equal to putting it in the operator?

Reply:
is it thought? It's my impression that tbb can reuse the same instance of the function on multiple ranges. If that's never the case then yes you're of course right but I'm pretty sure I recall reading somewhere that tbb is clever about when it is required to copy construct an instance and when it can be re-used (for tbb::parallel_for)

Reply:
There seems to be some uncertainty here, so can we agree that the worst case scenario of your suggested change is less efficiency whereas the worst case scenario of mine is a few extra lines of code? If so I think we should keep it as is. I'm fine revisiting this again once we're certain the change doesn't reduce performance, but as of now I'm not convinced.

Reply:
Yes that's absolutely fine